File: clients/go/connection.go
package sixsevendb

import (
	"context"
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"strconv"
	"strings"
	"sync"
	"time"
)

// rawConn is a low-level TCP connection to SixSevenDB using the PG wire protocol.
type rawConn struct {
	conn       net.Conn
	mu         sync.Mutex
	closed     bool
	parameters map[string]string
	pid        uint32
	secretKey  uint32
}

// dialConfig holds connection parameters.
type dialConfig struct {
	Host     string
	Port     int
	User     string
	Password string
	Database string
}

// defaultDialConfig returns defaults.
func defaultDialConfig() *dialConfig {
	return &dialConfig{
		Host:     "localhost",
		Port:     6767,
		User:     "sixseven",
		Database: "sixseven",
	}
}

// dial creates a new raw connection and performs the startup handshake.
func dial(ctx context.Context, cfg *dialConfig) (*rawConn, error) {
	addr := fmt.Sprintf("%s:%d", cfg.Host, cfg.Port)

	var d net.Dialer
	nc, err := d.DialContext(ctx, "tcp", addr)
	if err != nil {
		return nil, fmt.Errorf("sixsevendb: connect to %s: %w", addr, err)
	}

	rc := &rawConn{
		conn:       nc,
		parameters: make(map[string]string),
	}

	// Apply context deadline to the connection during startup.
	if deadline, ok := ctx.Deadline(); ok {
		_ = nc.SetDeadline(deadline)
		defer func() { _ = nc.SetDeadline(time.Time{}) }()
	}

	if err := rc.startup(cfg); err != nil {
		_ = nc.Close()
		return nil, err
	}

	return rc, nil
}

// startup performs the initial handshake and authentication.
func (c *rawConn) startup(cfg *dialConfig) error {
	if _, err := c.conn.Write(buildStartupMessage(cfg.User, cfg.Database)); err != nil {
		return fmt.Errorf("sixsevendb: send startup: %w", err)
	}
	return c.handleStartup(cfg)
}

// handleStartup reads messages until ReadyForQuery.
func (c *rawConn) handleStartup(cfg *dialConfig) error {
	for {
		msgType, payload, err := c.readMessage()
		if err != nil {
			return err
		}

		switch msgType {
		case msgAuthentication:
			if err := c.handleAuth(payload, cfg); err != nil {
				return err
			}
		case msgParameterStatus:
			name, nextPos := parseCString(payload, 0)
			value, _ := parseCString(payload, nextPos)
			c.parameters[name] = value
		case msgBackendKeyData:
			c.pid = binary.BigEndian.Uint32(payload[0:4])
			c.secretKey = binary.BigEndian.Uint32(payload[4:8])
		case msgReadyForQuery:
			return nil
		case msgErrorResponse:
			fields := parseErrorFields(payload)
			return &Error{
				Severity: fields['S'],
				Code:     fields['C'],
				Message:  fields['M'],
				Detail:   fields['D'],
				Hint:     fields['H'],
			}
		case msgNoticeResponse:
			continue
		}
	}
}

// handleAuth handles authentication challenge messages.
func (c *rawConn) handleAuth(payload []byte, cfg *dialConfig) error {
	if len(payload) < 4 {
		return fmt.Errorf("sixsevendb: authentication payload too short")
	}
	authType := binary.BigEndian.Uint32(payload[0:4])

	switch authType {
	case authOK:
		return nil
	case authCleartextPassword:
		if cfg.Password == "" {
			return fmt.Errorf("sixsevendb: server requires password but none provided")
		}
		_, err := c.conn.Write(buildPasswordMessage(cfg.Password))
		return err
	case authMD5Password:
		if cfg.Password == "" {
			return fmt.Errorf("sixsevendb: server requires password but none provided")
		}
		salt := payload[4:8]
		md5pw := buildMD5Password(cfg.User, cfg.Password, salt)
		_, err := c.conn.Write(buildPasswordMessage(md5pw))
		return err
	case authSASL:
		return c.handleSASL(payload[4:], cfg)
	default:
		return fmt.Errorf("sixsevendb: unsupported auth type %d", authType)
	}
}

// handleSASL performs the SCRAM-SHA-256 handshake.
func (c *rawConn) handleSASL(payload []byte, cfg *dialConfig) error {
	// Parse mechanisms
	var mechanisms []string
	pos := 0
	for pos < len(payload) {
		end := pos
		for end < len(payload) && payload[end] != 0 {
			end++
		}
		name := string(payload[pos:end])
		pos = end + 1
		if name == "" {
			break
		}
		mechanisms = append(mechanisms, name)
	}

	found := false
	for _, m := range mechanisms {
		if m == "SCRAM-SHA-256" {
			found = true
			break
		}
	}
	if !found {
		return fmt.Errorf("sixsevendb: server requires unsupported SASL mechanisms: %v", mechanisms)
	}
	if cfg.Password == "" {
		return fmt.Errorf("sixsevendb: server requires password but none provided")
	}

	state, clientFirst := scramClientFirst(cfg.User, cfg.Password)
	if _, err := c.conn.Write(buildSASLInitialResponse("SCRAM-SHA-256", clientFirst)); err != nil {
		return err
	}

	// Read SASLContinue
	contType, contPayload, err := c.readMessage()
	if err != nil {
		return err
	}
	if contType != msgAuthentication {
		return fmt.Errorf("sixsevendb: expected SASLContinue, got %c", contType)
	}
	if len(contPayload) < 4 || binary.BigEndian.Uint32(contPayload[0:4]) != authSASLContinue {
		return fmt.Errorf("sixsevendb: expected SASLContinue auth type")
	}

	clientFinal, err := scramClientFinal(state, contPayload[4:])
	if err != nil {
		return err
	}
	if _, err := c.conn.Write(buildSASLResponse(clientFinal)); err != nil {
		return err
	}

	// Read SASLFinal
	finalType, finalPayload, err := c.readMessage()
	if err != nil {
		return err
	}
	if finalType != msgAuthentication {
		return fmt.Errorf("sixsevendb: expected SASLFinal, got %c", finalType)
	}
	if len(finalPayload) < 4 || binary.BigEndian.Uint32(finalPayload[0:4]) != authSASLFinal {
		return fmt.Errorf("sixsevendb: expected SASLFinal auth type")
	}

	if !scramVerifyServer(state, finalPayload[4:]) {
		return fmt.Errorf("sixsevendb: server signature verification failed")
	}

	return nil
}

// readMessage reads one complete backend message from the connection.
func (c *rawConn) readMessage() (byte, []byte, error) {
	header := make([]byte, 5)
	if _, err := io.ReadFull(c.conn, header); err != nil {
		return 0, nil, fmt.Errorf("sixsevendb: read message header: %w", err)
	}
	msgType := header[0]
	length := binary.BigEndian.Uint32(header[1:5])
	if length < 4 {
		return 0, nil, fmt.Errorf("sixsevendb: invalid message length %d", length)
	}
	payloadLen := length - 4
	payload := make([]byte, payloadLen)
	if payloadLen > 0 {
		if _, err := io.ReadFull(c.conn, payload); err != nil {
			return 0, nil, fmt.Errorf("sixsevendb: read message payload: %w", err)
		}
	}
	return msgType, payload, nil
}

// writeMessage writes raw bytes to the connection.
func (c *rawConn) writeMessage(data []byte) error {
	_, err := c.conn.Write(data)
	return err
}

// simpleQuery executes a query via the simple query protocol (no parameters).
func (c *rawConn) simpleQuery(ctx context.Context, sql string) ([]fieldDescription, [][][]byte, string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil, nil, "", ErrClosed
	}

	// Set deadline from context
	if deadline, ok := ctx.Deadline(); ok {
		_ = c.conn.SetDeadline(deadline)
		defer func() { _ = c.conn.SetDeadline(time.Time{}) }()
	}

	if err := c.writeMessage(buildQueryMessage(sql)); err != nil {
		return nil, nil, "", err
	}

	var fields []fieldDescription
	var rows [][][]byte
	var command string

	for {
		msgType, payload, err := c.readMessage()
		if err != nil {
			return nil, nil, "", err
		}

		switch msgType {
		case msgRowDescription:
			fields, err = parseRowDescription(payload)
			if err != nil {
				return nil, nil, "", err
			}
		case msgDataRow:
			row, err := parseDataRow(payload)
			if err != nil {
				return nil, nil, "", err
			}
			rows = append(rows, row)
		case msgCommandComplete:
			command = parseCStringSimple(payload)
		case msgEmptyQueryResp:
			// empty query
		case msgErrorResponse:
			errFields := parseErrorFields(payload)
			serverErr := &Error{
				Severity: errFields['S'],
				Code:     errFields['C'],
				Message:  errFields['M'],
				Detail:   errFields['D'],
				Hint:     errFields['H'],
			}
			// Wait for ReadyForQuery before returning the error
			c.waitForReady()
			return nil, nil, "", serverErr
		case msgNoticeResponse:
			continue
		case msgReadyForQuery:
			return fields, rows, command, nil
		}
	}
}

// extendedQuery executes a parameterized query via the extended query protocol.
func (c *rawConn) extendedQuery(ctx context.Context, sql string, args []string) ([]fieldDescription, [][][]byte, string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil, nil, "", ErrClosed
	}

	if deadline, ok := ctx.Deadline(); ok {
		_ = c.conn.SetDeadline(deadline)
		defer func() { _ = c.conn.SetDeadline(time.Time{}) }()
	}

	// Send Parse + Bind + Describe + Execute + Sync
	var buf []byte
	buf = append(buf, buildParseMessage(sql, "")...)
	buf = append(buf, buildBindMessage(args, "", "")...)
	buf = append(buf, buildDescribeMessage('P', "")...)
	buf = append(buf, buildExecuteMessage("", 0)...)
	buf = append(buf, buildSyncMessage()...)

	if err := c.writeMessage(buf); err != nil {
		return nil, nil, "", err
	}

	var fields []fieldDescription
	var rows [][][]byte
	var command string

	for {
		msgType, payload, err := c.readMessage()
		if err != nil {
			return nil, nil, "", err
		}

		switch msgType {
		case msgParseComplete, msgBindComplete, msgNoData:
			continue
		case msgRowDescription:
			fields, err = parseRowDescription(payload)
			if err != nil {
				return nil, nil, "", err
			}
		case msgDataRow:
			row, err := parseDataRow(payload)
			if err != nil {
				return nil, nil, "", err
			}
			rows = append(rows, row)
		case msgCommandComplete:
			command = parseCStringSimple(payload)
		case msgEmptyQueryResp:
			// empty query
		case msgErrorResponse:
			errFields := parseErrorFields(payload)
			serverErr := &Error{
				Severity: errFields['S'],
				Code:     errFields['C'],
				Message:  errFields['M'],
				Detail:   errFields['D'],
				Hint:     errFields['H'],
			}
			c.waitForReady()
			return nil, nil, "", serverErr
		case msgNoticeResponse:
			continue
		case msgReadyForQuery:
			return fields, rows, command, nil
		}
	}
}

// waitForReady reads messages until ReadyForQuery is received.
func (c *rawConn) waitForReady() {
	for {
		msgType, _, err := c.readMessage()
		if err != nil || msgType == msgReadyForQuery {
			return
		}
	}
}

// close closes the connection gracefully.
func (c *rawConn) close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil
	}
	c.closed = true

	_ = c.writeMessage(buildTerminateMessage())
	return c.conn.Close()
}

// parseCStringSimple extracts a null-terminated string from the start of payload.
func parseCStringSimple(payload []byte) string {
	end := 0
	for end < len(payload) && payload[end] != 0 {
		end++
	}
	return string(payload[:end])
}

// parseCommandTag extracts the command name from a CommandComplete tag.
func parseCommandTag(tag string) string {
	parts := strings.SplitN(tag, " ", 2)
	if len(parts) > 0 {
		return parts[0]
	}
	return ""
}

// parseRowCount extracts the row count from a CommandComplete tag.
func parseRowCount(tag string) int64 {
	parts := strings.Split(tag, " ")
	if len(parts) >= 2 {
		count, err := strconv.ParseInt(parts[len(parts)-1], 10, 64)
		if err == nil {
			return count
		}
	}
	return 0
}

File: clients/rust/src/connection.rs
use std::collections::HashMap;

use tokio::net::TcpStream;

use crate::config::Config;
use crate::error::{Error, Result};
use crate::protocol::*;
use crate::types::{self, Value};

/// A raw connection to a SixSevenDB server.
pub struct Connection {
    stream: TcpStream,
    closed: bool,
    /// Server parameters received during startup (e.g., "server_version").
    pub parameters: HashMap<String, String>,
    /// Backend process ID.
    pub pid: u32,
    /// Cancel secret key.
    pub secret_key: u32,
}

/// Query result from executing a statement.
#[derive(Debug)]
pub struct QueryResult {
    /// Column descriptions.
    pub fields: Vec<FieldDescription>,
    /// Rows of data (each row is a vector of optional byte vectors).
    pub rows: Vec<Vec<Option<Vec<u8>>>>,
    /// Command completion tag (e.g., "SELECT 5").
    pub command_tag: String,
}

impl QueryResult {
    /// Number of rows affected (parsed from command tag).
    pub fn rows_affected(&self) -> u64 {
        parse_row_count(&self.command_tag)
    }

    /// Parse all rows into typed Values.
    pub fn typed_rows(&self) -> Vec<Vec<Value>> {
        self.rows
            .iter()
            .map(|row| {
                row.iter()
                    .enumerate()
                    .map(|(i, col)| {
                        match col {
                            None => Value::Null,
                            Some(bytes) => {
                                let text = String::from_utf8_lossy(bytes);
                                let oid = self.fields.get(i).map(|f| f.type_oid).unwrap_or(0);
                                types::parse_value(oid, &text).unwrap_or(Value::Text(text.into_owned()))
                            }
                        }
                    })
                    .collect()
            })
            .collect()
    }
}

impl Connection {
    /// Connect to a SixSevenDB server using the given config.
    pub async fn connect(cfg: &Config) -> Result<Self> {
        let stream = TcpStream::connect(cfg.address()).await?;
        let mut conn = Connection {
            stream,
            closed: false,
            parameters: HashMap::new(),
            pid: 0,
            secret_key: 0,
        };

        conn.startup(cfg).await?;
        Ok(conn)
    }

    /// Connect using a DSN string.
    pub async fn connect_dsn(dsn: &str) -> Result<Self> {
        let cfg = crate::config::parse_dsn(dsn)?;
        Self::connect(&cfg).await
    }

    /// Send startup message and handle handshake.
    async fn startup(&mut self, cfg: &Config) -> Result<()> {
        let msg = build_startup_message(&cfg.user, &cfg.database);
        self.stream.try_write(&msg).map_err(Error::Io)?;
        // Flush after writing startup message
        tokio::io::AsyncWriteExt::flush(&mut self.stream).await?;
        self.handle_startup(cfg).await
    }

    /// Handle the startup handshake messages until ReadyForQuery.
    async fn handle_startup(&mut self, cfg: &Config) -> Result<()> {
        loop {
            let msg = read_message(&mut self.stream).await?;
            match msg.msg_type {
                MSG_AUTH => {
                    if msg.payload.len() < 4 {
                        return Err(Error::Protocol("auth payload too short".into()));
                    }
                    let auth_type = u32::from_be_bytes([
                        msg.payload[0], msg.payload[1], msg.payload[2], msg.payload[3],
                    ]);
                    if auth_type == AUTH_OK {
                        continue;
                    }
                    crate::auth::handle_auth(
                        &mut self.stream,
                        auth_type,
                        &msg.payload,
                        &cfg.user,
                        &cfg.password,
                    )
                    .await?;
                }
                MSG_PARAMETER_STATUS => {
                    let (key, next) = parse_cstring(&msg.payload, 0);
                    let (value, _) = parse_cstring(&msg.payload, next);
                    self.parameters.insert(key, value);
                }
                MSG_BACKEND_KEY_DATA => {
                    if msg.payload.len() >= 8 {
                        self.pid = u32::from_be_bytes([
                            msg.payload[0], msg.payload[1], msg.payload[2], msg.payload[3],
                        ]);
                        self.secret_key = u32::from_be_bytes([
                            msg.payload[4], msg.payload[5], msg.payload[6], msg.payload[7],
                        ]);
                    }
                }
                MSG_READY_FOR_QUERY => {
                    return Ok(());
                }
                MSG_ERROR_RESPONSE => {
                    let err = parse_error_fields(&msg.payload);
                    return Err(err.into());
                }
                MSG_NOTICE_RESPONSE => {
                    // Ignored
                }
                _ => {
                    tracing::debug!("ignoring startup message type: {}", msg.msg_type as char);
                }
            }
        }
    }

    /// Execute a simple query (no parameters).
    pub async fn simple_query(&mut self, sql: &str) -> Result<QueryResult> {
        self.check_closed()?;

        let payload = build_query_message(sql);
        write_message(&mut self.stream, b'Q', &payload).await?;

        self.read_query_result().await
    }

    /// Execute a parameterized query using the extended protocol.
    pub async fn query(&mut self, sql: &str, params: &[Value]) -> Result<QueryResult> {
        self.check_closed()?;

        if params.is_empty() {
            return self.simple_query(sql).await;
        }

        // Convert params to text format
        let text_params: Vec<Option<String>> = params
            .iter()
            .map(types::format_value)
            .collect();

        // Send Parse + Bind + Describe + Execute + Sync
        let parse = build_parse_message(sql, "");
        write_message(&mut self.stream, b'P', &parse).await?;

        let bind = build_bind_message(&text_params, "", "");
        write_message(&mut self.stream, b'B', &bind).await?;

        let describe = build_describe_message(b'P', "");
        write_message(&mut self.stream, b'D', &describe).await?;

        let execute = build_execute_message("", 0);
        write_message(&mut self.stream, b'E', &execute).await?;

        let sync = build_sync_message();
        write_message(&mut self.stream, b'S', &sync).await?;

        self.read_extended_result().await
    }

    /// Execute a statement and return only the number of rows affected.
    pub async fn execute(&mut self, sql: &str, params: &[Value]) -> Result<u64> {
        let result = self.query(sql, params).await?;
        Ok(result.rows_affected())
    }

    /// Ping the server with SELECT 1.
    pub async fn ping(&mut self) -> Result<()> {
        self.simple_query("SELECT 1").await?;
        Ok(())
    }

    /// Close the connection.
    pub async fn close(&mut self) -> Result<()> {
        if self.closed {
            return Ok(());
        }
        let payload = build_terminate_message();
        let _ = write_message(&mut self.stream, b'X', &payload).await;
        self.closed = true;
        Ok(())
    }

    /// Check if the connection is closed.
    pub fn is_closed(&self) -> bool {
        self.closed
    }

    fn check_closed(&self) -> Result<()> {
        if self.closed {
            Err(Error::Closed)
        } else {
            Ok(())
        }
    }

    /// Read result from a simple query.
    async fn read_query_result(&mut self) -> Result<QueryResult> {
        let mut fields = Vec::new();
        let mut rows = Vec::new();
        let mut command_tag = String::new();

        loop {
            let msg = read_message(&mut self.stream).await?;
            match msg.msg_type {
                MSG_ROW_DESCRIPTION => {
                    fields = parse_row_description(&msg.payload);
                }
                MSG_DATA_ROW => {
                    rows.push(parse_data_row(&msg.payload));
                }
                MSG_COMMAND_COMPLETE => {
                    let (tag, _) = parse_cstring(&msg.payload, 0);
                    command_tag = tag;
                }
                MSG_EMPTY_QUERY => {}
                MSG_ERROR_RESPONSE => {
                    let err = parse_error_fields(&msg.payload);
                    // Drain until ReadyForQuery
                    self.wait_for_ready().await?;
                    return Err(err.into());
                }
                MSG_NOTICE_RESPONSE => {}
                MSG_READY_FOR_QUERY => {
                    return Ok(QueryResult { fields, rows, command_tag });
                }
                _ => {
                    tracing::debug!("ignoring message type in query: {}", msg.msg_type as char);
                }
            }
        }
    }

    /// Read result from an extended protocol query.
    async fn read_extended_result(&mut self) -> Result<QueryResult> {
        let mut fields = Vec::new();
        let mut rows = Vec::new();
        let mut command_tag = String::new();

        loop {
            let msg = read_message(&mut self.stream).await?;
            match msg.msg_type {
                MSG_PARSE_COMPLETE | MSG_BIND_COMPLETE | MSG_NO_DATA => {}
                MSG_ROW_DESCRIPTION => {
                    fields = parse_row_description(&msg.payload);
                }
                MSG_DATA_ROW => {
                    rows.push(parse_data_row(&msg.payload));
                }
                MSG_COMMAND_COMPLETE => {
                    let (tag, _) = parse_cstring(&msg.payload, 0);
                    command_tag = tag;
                }
                MSG_EMPTY_QUERY => {}
                MSG_ERROR_RESPONSE => {
                    let err = parse_error_fields(&msg.payload);
                    self.wait_for_ready().await?;
                    return Err(err.into());
                }
                MSG_NOTICE_RESPONSE => {}
                MSG_READY_FOR_QUERY => {
                    return Ok(QueryResult { fields, rows, command_tag });
                }
                _ => {
                    tracing::debug!("ignoring extended message type: {}", msg.msg_type as char);
                }
            }
        }
    }

    /// Drain messages until ReadyForQuery.
    async fn wait_for_ready(&mut self) -> Result<()> {
        loop {
            let msg = read_message(&mut self.stream).await?;
            if msg.msg_type == MSG_READY_FOR_QUERY {
                return Ok(());
            }
        }
    }

}

impl Drop for Connection {
    fn drop(&mut self) {
        if !self.closed {
            self.closed = true;
            // Best-effort close; async close isn't possible in Drop
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_query_result_rows_affected() {
        let result = QueryResult {
            fields: vec![],
            rows: vec![],
            command_tag: "INSERT 0 5".to_string(),
        };
        assert_eq!(result.rows_affected(), 5);

        let result2 = QueryResult {
            fields: vec![],
            rows: vec![],
            command_tag: "DELETE 3".to_string(),
        };
        assert_eq!(result2.rows_affected(), 3);
    }

    #[test]
    fn test_query_result_typed_rows() {
        let result = QueryResult {
            fields: vec![
                FieldDescription {
                    name: "id".into(),
                    table_oid: 0,
                    column_index: 0,
                    type_oid: 23, // INT4
                    type_size: 4,
                    type_modifier: -1,
                    format_code: 0,
                },
                FieldDescription {
                    name: "name".into(),
                    table_oid: 0,
                    column_index: 1,
                    type_oid: 25, // TEXT
                    type_size: -1,
                    type_modifier: -1,
                    format_code: 0,
                },
            ],
            rows: vec![
                vec![Some(b"42".to_vec()), Some(b"Alice".to_vec())],
                vec![Some(b"7".to_vec()), None],
            ],
            command_tag: "SELECT 2".to_string(),
        };

        let typed = result.typed_rows();
        assert_eq!(typed.len(), 2);
        assert_eq!(typed[0][0], Value::Int4(42));
        assert_eq!(typed[0][1], Value::Text("Alice".into()));
        assert_eq!(typed[1][0], Value::Int4(7));
        assert_eq!(typed[1][1], Value::Null);
    }
}
